docs/field_write_semantics.adoc

```diff
@@ -0,0 +1,129 @@
+
+== Field Write Semantics
+
+Each field in a register is parameterized with a position -- the closed
+(inclusive) range from LSB to MSB, from which a mask can be derived -- and a
+_write function_.
+
+=== Write functions
+
+Examples of write functions are:
+
+- replacement ("normal write")
+- read-only
+- write 1 to clear
+
+There are 16 possible write functions, each characterized by a truth table
+relating the current value *C*, the written value *W* and the result *R*. For
+example:
+
+.Truth table for replacement
+[width=30%,cols="1,1,1"]
+|===
+^h|C ^h|W ^h|R
+
+^|0 ^|0 ^|0
+^|0 ^|1 ^|1
+^|1 ^|0 ^|0
+^|1 ^|1 ^|1
+|===
+
+.Truth table for read-only
+[width=30%,cols="1,1,1"]
+|===
+^h|C ^h|W ^h|R
+
+^|0 ^|0 ^|0
+^|0 ^|1 ^|0
+^|1 ^|0 ^|1
+^|1 ^|1 ^|1
+|===
+
+=== Identity specs
+
+Each such write function has an _identity spec_ indicated by its truth table.
+For replacement, the identity spec is *none*. For read-only, the identity spec
+is *don't care*. There are 5 possible identity specs, denoted as follows:
+
+- *E* - none
+- *0* - 0 is the identity
+- *1* - 1 is the identity
+- *X* - don't care
+- *** - either 0 or 1 is the identity
+
+NOTE: *\** is not a value associated with any write function, but it is necessary
+for this operation to be closed. The difference between *X* and *** is that *X*
+is truly "don't care" where *\** is one of *0* or *1*, and only one of them
+works. *** could also be termed the "natural" or "dependent" value for the
+identity.
+
+We can define a binary commutative function on these identity specs:
+
+.Identity spec function
+[width=30%,cols="1,1,1"]
+|===
+^h|A ^h|B ^h|R
+
+^|E ^|E ^|E
+^|E ^|0 ^|E
+^|E ^|1 ^|E
+^|E ^|X ^|E
+^|E ^|* ^|E
+
+^|0 ^|0 ^|0
+^|0 ^|1 ^|*
+^|0 ^|X ^|0
+^|0 ^|* ^|*
+
+^|1 ^|1 ^|1
+^|1 ^|X ^|1
+^|1 ^|* ^|*
+
+^|X ^|* ^|*
+^|X ^|X ^|X
+
+^|* ^|* ^|*
+|===
+
+=== Building the masks
+
+In general a given write to a register will only want to write some of the field
+values. The other field values are left unspecified. We can compute the common
+identity spec for the write by folding the spec function across all the fields
```

Comment:
Question: do we need to find a common identity spec across all unspecified fields? i believe we can evaluate the identity function for each field on its own to determine if a RMW is necessary or we can just write 1 or 0.

Reply:
I think we're saying the same thing. The identity spec function _is_ how we find out whether RMW is necessary. The identity for a field isn't a function, it's a value. This is very general: it allows for fields in a register to have different write functions.

Reply:
Bear in mind that "all the fields across the register" could also mean "all the fields across the addressable unit" if we combine this with byte enables.